src/Features/CSharp/Portable/RemoveUnusedParametersAndValues/CSharpRemoveUnusedParametersAndValuesCodeFixProvider.cs

```diff
@@ -0,0 +1,187 @@
+    {
+        protected override BlockSyntax GenerateBlock(IEnumerable<StatementSyntax> statements)
+            => SyntaxFactory.Block(statements);
+
+        protected override SyntaxToken GetForEachStatementIdentifier(ForEachStatementSyntax node)
+            => node.Identifier;
+
+        protected override SyntaxNode UpdateNameForFlaggedNode(SyntaxNode node, SyntaxToken newName)
+        {
+            switch (node.Kind())
+            {
+                case SyntaxKind.IdentifierName:
+                    var identifierName = (IdentifierNameSyntax)node;
+                    return identifierName.WithIdentifier(newName.WithTriviaFrom(identifierName.Identifier));
+
+                case SyntaxKind.VariableDeclarator:
+                    var variableDeclarator = (VariableDeclaratorSyntax)node;
+                    return variableDeclarator.WithIdentifier(newName.WithTriviaFrom(variableDeclarator.Identifier));
+
+                case SyntaxKind.SingleVariableDesignation:
+                    return SyntaxFactory.SingleVariableDesignation(newName).WithTriviaFrom(node);
+
+                case SyntaxKind.CatchDeclaration:
+                    var catchDeclaration = (CatchDeclarationSyntax)node;
+                    return catchDeclaration.WithIdentifier(newName.WithTriviaFrom(catchDeclaration.Identifier));
+
+                default:
+                    throw ExceptionUtilities.Unreachable;
+            }
+        }
+
+        protected override ILocalSymbol GetSingleDeclaredLocal(LocalDeclarationStatementSyntax localDeclaration, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            Contract.ThrowIfFalse(localDeclaration.Declaration.Variables.Count == 1);
+            return (ILocalSymbol)semanticModel.GetDeclaredSymbol(localDeclaration.Declaration.Variables[0]);
+        }
+
+        protected override void InsertAtStartOfSwitchCaseBlock(SwitchSectionSyntax switchCaseBlock, SyntaxEditor editor, LocalDeclarationStatementSyntax declarationStatement)
+        {
+            var firstStatement = switchCaseBlock.Statements.FirstOrDefault();
+            if (firstStatement != null)
+            {
+                editor.InsertBefore(firstStatement, declarationStatement);
+            }
+            else
+            {
+                // Switch section without any statements is an error case.
+                // Insert before containing switch statement.
+                editor.InsertBefore(switchCaseBlock.Parent, declarationStatement);
+            }
+        }
+
+        protected override Task<SyntaxNode> RemoveDiscardDeclarationsAsync(
+            SyntaxNode memberDeclaration,
+            Document document,
+            CancellationToken cancellationToken)
+        {
+            var editor = new SyntaxEditor(memberDeclaration, SyntaxGenerator.GetGenerator(document));
+            foreach (var child in memberDeclaration.DescendantNodes(descendIntoChildren: n => !(n is ExpressionSyntax)))
```

Comment:
comment as to why you're not descending into expressions.  for example, what if you have a discard inside a lambda somewhere?

Reply:
Thanks, I hadn't considered that case and tried to optimize the tree traversal here. I have removed this check and also added unit tests to verify the same.

---
In reply to: [230603284](https://github.com/dotnet/roslyn/pull/30777#discussion_r230603284) [](ancestors = 230603284)